gitpod-io/gitpod
# Proper webhook installation detection

## Description

Improves the way we detect if you have webhooks installed on a specific repo. The approach with this PR changes to the following:
1. We check the host if it knows about a webhook for the current Gitpod host. If we lack permissions, we continue to step 2, otherwise, we return the result of this check.
2. (GitHub-specific) we look at the `pushedAt` date to determine when we should've received the last push event and see if we have received some webhook event after that date. If so, we return it.
3. We query the latest webhook events we got for the repo. We go through them one-by-one and if one of the events matches the latest commit SHA of the repo's default branch, we return that event.

## Related Issues

Contributes to ENT-601 as a follow-up for https://github.com/gitpod-io/gitpod/pull/20228

## How to test


https://ft-super-d2d2fa388b8.preview.gitpod-dev.com/repositories

1. Import a repository
2. Enable prebuilds on it (should install webhook)
3. The prebuild settings page should say that the trigger strategy is "Webhook-based"
4. Remove the webhook on the Git host
5. Create a workspace for that repository, wait for it to create
6. Check its prebuild settings again, it should now say "Activity-based" in the trigger strategy section.

/hold
